fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/IndexScrubbing.java

```diff
@@ -0,0 +1,222 @@
+                rangeSet.clear();
+                return AsyncUtil.READY_FALSE;
+            }
+            final Tuple rangeStart = RangeSet.isFirstKey(range.begin) ? null : Tuple.fromBytes(range.begin);
+            final Tuple rangeEnd = RangeSet.isFinalKey(range.end) ? null : Tuple.fromBytes(range.end);
+            final TupleRange tupleRange = TupleRange.between(rangeStart, rangeEnd);
+
+            final RecordCursor<T> cursor = tools.getIterator(tupleRange, store, getLimit() + 1); // always respect limit in this path; +1 allows a continuation item in forward scan
+            final AtomicBoolean hasMore = new AtomicBoolean(true);
+            final AtomicReference<RecordCursorResult<T>> lastResult = new AtomicReference<>(RecordCursorResult.exhausted());
+            final long scanLimit = scrubbingPolicy.getEntriesScanLimit();
+            List<IndexScrubbingTools.Issue> issueList = new LinkedList<>();
+
+            return iterateRangeOnly(store, cursor, (recordStore, result) -> handleOneItem(recordStore, result, tools, issueList),
+                    lastResult, hasMore, recordsScanned, isIdempotent)
+                    .thenApply(vignore -> hasMore.get() ? tools.getContinuation(lastResult.get()) :
+                                          rangeEnd)
+                    .thenCompose(cont -> rangeSet.insertRangeAsync(packOrNull(rangeStart), packOrNull(cont), true)
+                            .thenApply(ignore -> notAllRangesExhausted(cont, rangeEnd)))
+                    .thenApply(ret -> {
+                        // Here: after a successful transaction, report the issues and check/update the scan counter.
+                        reportIssues(issueList);
+                        if (scanLimit > 0) {
+                            scanCounter += recordsScanned.get();
+                            if (scanLimit <= scanCounter) {
+                                return false;
+                            }
+                        }
+                        return ret;
+                    });
+        });
+    }
+
+
+    private <T> CompletableFuture<FDBStoredRecord<Message>> handleOneItem(FDBRecordStore store, final RecordCursorResult<T> result, final IndexScrubbingTools<T> tools, List<IndexScrubbingTools.Issue> issueList) {
+        return tools.handleOneItem(store,
+                        store.getIndexMaintainer(common.getIndex()).state.transaction,
+                        result)
+                .thenApply(issue -> {
+                    if (issue == null) {
+                        return null;
+                    }
+                    issueList.add(issue);
+                    return issue.recordToIndex;
+                });
+    }
+
+    private void reportIssues(List<IndexScrubbingTools.Issue> issueList) {
+        // report these issues only after their transaction was completed successfully
+        for (IndexScrubbingTools.Issue issue: issueList) {
```

Comment:
**Forward thinking**
To improve testing, having the `IndexingScrubber` optionally take a `Consumer<Issue>` instead of just logging, the tests could assert that they see specific issues come up, and the issues are what was expected.